File: sunflower-core/src/main/java/org/tdf/sunflower/state/AccountTrie.java
package org.tdf.sunflower.state;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections4.ListUtils;
import org.tdf.common.serialize.Codecs;
import org.tdf.common.store.NoDeleteStore;
import org.tdf.common.store.Store;
import org.tdf.common.trie.SecureTrie;
import org.tdf.common.trie.Trie;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.Start;
import org.tdf.sunflower.types.CryptoContext;
import org.tdf.sunflower.types.Header;
import org.tdf.sunflower.vm.Backend;
import org.tdf.sunflower.vm.BackendImpl;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Getter
@Slf4j(topic = "trie")
@AllArgsConstructor
public class AccountTrie extends AbstractStateTrie<HexBytes, Account> {
    private Trie<HexBytes, Account> trie;
    private final Trie<HexBytes, HexBytes> contractStorageTrie;
    private final Store<HexBytes, HexBytes> contractCodeStore;
    private Map<HexBytes, BuiltinContract> bios;
    private Map<HexBytes, BuiltinContract> builtins;
    private final Store<byte[], byte[]> db;
    private final Store<byte[], byte[]> trieStore;

    public Backend createBackend(
        Header parent,
        HexBytes root,
        Long newBlockCreatedAt,
        boolean isStatic
    ) {
        return new BackendImpl(
            parent,
            null,
            trie.revert(root),
            contractStorageTrie,
            new HashMap<>(),
            new HashMap<>(),
            builtins,
            bios,
            isStatic,
            contractCodeStore,
            new HashMap<>(),
            newBlockCreatedAt
        );
    }

    @SneakyThrows
    public AccountTrie(
        Store<byte[], byte[]> db,
        Store<HexBytes, HexBytes> contractCodeStore,
        Trie<HexBytes, HexBytes> contractStorageTrie,
        boolean secure
    ) {
        this.db = db;
        this.trieStore = new NoDeleteStore<>(db, Store.IS_NULL);

        this.trie = Trie.<HexBytes, Account>builder()
            .hashFunction(CryptoContext::hash)
            .store(trieStore)
            .keyCodec(Codecs.newRLPCodec(HexBytes.class))
            .valueCodec(Codecs.newRLPCodec(Account.class))
            .build();

        if (secure)
            trie = new SecureTrie<>(trie, CryptoContext::hash);

        this.contractStorageTrie = contractStorageTrie;
        this.contractCodeStore = contractCodeStore;
    }

    @Override
    @SneakyThrows
    public HexBytes init(List<Account> alloc, List<BuiltinContract> bios, List<BuiltinContract> builtins) {
        this.builtins = new HashMap<>();
        for (BuiltinContract builtin : builtins) {
            this.builtins.put(builtin.getAddress(), builtin);
        }

        this.bios = new HashMap<>();
        for (BuiltinContract b : bios) {
            this.bios.put(b.getAddress(), b);
        }

        Map<HexBytes, Account> genesisStates = new HashMap<>();

        for (Account account : alloc) {
            genesisStates.put(account.getAddress(), account);
        }


        for (BuiltinContract c : ListUtils.sum(bios, builtins)) {
            HexBytes address = c.getAddress();
            Account a = Account.emptyAccount(address, Uint256.ZERO);
            Trie<HexBytes, HexBytes> trie = contractStorageTrie.revert();
            for (Map.Entry<HexBytes, HexBytes> entry : c.getGenesisStorage().entrySet()) {
                trie.set(entry.getKey(), entry.getValue());
            }
            HexBytes root = trie.commit();
            trie.flush();
            a.setStorageRoot(root);
            genesisStates.put(address, a);
        }

        // sync to genesis
        Trie<HexBytes, Account> tmp = this.trie.revert();
        genesisStates.forEach(tmp::set);
        log.info("genesis states = {}", Start.MAPPER.writeValueAsString(genesisStates));
        HexBytes r = tmp.commit();
        log.info("genesis state root = " + r);
        tmp.flush();
        return r;
    }

    @Override
    Store<byte[], byte[]> getDB() {
        return this.db;
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/state/AccountTrie.kt
package org.tdf.sunflower.state

import org.slf4j.LoggerFactory
import org.tdf.common.serialize.Codecs
import org.tdf.common.store.NoDeleteStore
import org.tdf.common.store.Store
import org.tdf.common.trie.SecureTrie
import org.tdf.common.trie.Trie
import org.tdf.common.trie.TrieImpl
import org.tdf.common.util.HexBytes
import org.tdf.sunflower.Start
import org.tdf.sunflower.facade.RepositoryReader
import org.tdf.sunflower.types.Header
import org.tdf.sunflower.vm.*
import org.tdf.sunflower.vm.hosts.Limit


class AccountTrie(
    override val db: Store<ByteArray, ByteArray>,
    private val contractCodeStore: Store<HexBytes, HexBytes>,
    val contractStorageTrie: Trie<HexBytes, HexBytes>,
    secure: Boolean
) : AbstractStateTrie<HexBytes, Account>() {
    override val trie: Trie<HexBytes, Account>

    lateinit var bios: Map<HexBytes, Builtin>
    lateinit var builtins: Map<HexBytes, Builtin>

    override val trieStore: Store<ByteArray, ByteArray>

    override fun createBackend(
        parent: Header,
        staticCall: Boolean,
        root: HexBytes,
    ): Backend {
        return BackendImpl(
            parent = parent,
            trie = trie.revert(root),
            contractStorageTrie = contractStorageTrie,
            builtins = builtins,
            bios = bios,
            staticCall = staticCall,
            codeStore = contractCodeStore,
        )
    }

    init {
        trieStore = NoDeleteStore(db) { it == null || it.isEmpty() }
        var trie: Trie<HexBytes, Account> = TrieImpl(
            trieStore,
            Codecs.rlp(HexBytes::class.java),
            Codecs.rlp(Account::class.java),
        )
        if (secure)
            trie = SecureTrie(trie)
        this.trie = trie
    }

    override fun init(
        alloc: Map<HexBytes, Account>,
        bios: List<Builtin>,
        builtins: List<Builtin>,
        consensusCode: Map<HexBytes, HexBytes>,
        rd: RepositoryReader
    ): HexBytes {
        this.builtins = builtins.associateBy { it.address }
        this.bios = bios.associateBy { it.address }

        val genesisStates: MutableMap<HexBytes, Account> = alloc.toMutableMap()

        for (c in (bios + builtins)) {
            if(c.pure) continue
            val address = c.address

            val trie = contractStorageTrie.revert()
            for ((key, value) in c.genesisStorage) {
                trie[key] = value
            }
            val root = trie.commit()
            trie.flush()
            val a = Account(storageRoot = root)
            genesisStates[address] = a
        }


        // sync to genesis
        val tmp = trie.revert()
        genesisStates.forEach { tmp[it.key] = it.value }
        tmp.commit()

        val backend = BackendImpl(
            parent = null,
            trie = tmp,
            contractStorageTrie = contractStorageTrie,
            builtins = this.builtins,
            bios = this.bios,
            staticCall = false,
            codeStore = contractCodeStore,
        )

        // create consensus code
        consensusCode.forEach { (t, u) ->
            log.info("deploy consensus code at address {}", t)
            val cd = CallData(to = t, callType = CallType.CREATE, data = u)
            val ex = VMExecutor(rd, backend, CallContext(), cd, Limit(VMExecutor.GAS_UNLIMITED), mutableListOf())
            ex.execute()
        }

        log.info("genesis states = {}", Start.MAPPER.writeValueAsString(genesisStates))
        val r = backend.merge()
        log.info("genesis state root = $r")
        return r
    }


    companion object {
        private val log = LoggerFactory.getLogger("trie")
    }
}